### PR TITLE
Improve docs on ContractStateMachine

### PR DESCRIPTION
**Pull request by cocreature**
Based on Matthias suggestions from https://github.com/digital-asset/daml/commit/cf6fdea0eb73c9017e377b1022993dce5ceeddf2

changelog_begin
changelog_end

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 1]
```diff
@@ -21,32 +21,8 @@ import com.daml.lf.value.Value.ContractId
...
-  * handled with [[ContractStateMachine.State.handleLookupWith]],
-  * whose second argument takes the contract key resolution to be given to the Daml interpreter instead of
-  * [[com.daml.lf.transaction.Node.LookupByKey.result]].
-  * This is because the iteration might currently be within a rollback scope
-  * that has already archived a contract with the key without a by-key operation
-  * (and for this reason the archival is not tracked in [[ContractStateMachine.ActiveLedgerState.keys]]),
-  * i.e., the lookup resolves to [[scala.None$]] but the correct key input is [[scala.Some$]] for some contract ID
-  * and this may matter after the rollback scope is left.
-  *
-  * Note that the engine is also used in Canton’s non-uck (unique contract key) mode.
```
Comment (matthiasS-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:47]: I liked this very last paragraph. I think we should have some disclaimer about non-uck mode.

Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:47]: the disclaimer is on the mode. I don’t think duplicating it adds much.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 2]
```diff
@@ -479,12 +508,12 @@ object ContractStateMachine {
     *        2.1. A create will set the corresponding map entry to KeyActive(cid) if the contract has a key.
     *        2.2. A consuming choice on cid will set the corresponding map entry to KeyInactive
```
Comment (matthiasS-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:509]:
```suggestion
    *        2.2. A consuming exerciseByKey on cid will set the corresponding map entry to KeyInactive
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 3]
```diff
@@ -479,12 +508,12 @@ object ContractStateMachine {
...
     *             iff we had a KeyActive(cid) entry for the same key before. If not, keys
-    *             will not be modified. Later lookups have an activeness check
+    *             will not be modified.
+    *             Later lookups have an activeness check
```
Comment (matthiasS-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:512]: I would try to merge this into Point 1, starting at L505. See https://github.com/digital-asset/daml/commit/cf6fdea0eb73c9017e377b1022993dce5ceeddf2#commitcomment-76159212

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 4]
```diff
@@ -216,13 +197,54 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
...
+      * [[com.daml.lf.transaction.Node.LookupByKey.result]].
+      * This is because the iteration might currently be within a rollback scope
+      * that has already archived a contract with the key without a by-key operation
+      * (and for this reason the archival is not tracked in [[ContractStateMachine.ActiveLedgerState.keys]]),
+      * i.e., the lookup resolves to [[scala.None$]] but the correct key input is [[scala.Some$]] for some contract ID
+      * and this may matter after the rollback scope is left.
+      */
+    def handleLookupWith(
+        lookup: Node.LookupByKey,
+        keyInput: Option[ContractId],
```
Comment (matthiasS-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:220]: I wonder how the caller is supposed to make up the correct value for `keyInput` from? Could you please provide some guidance?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala [hunk 5]
```diff
@@ -216,13 +197,54 @@ class ContractStateMachine[Nid](mode: ContractKeyUniquenessMode) {
...
+    ): Either[KeyInputError, State] = {
+      require(mode.byKeyOnly, "This method can only be used if only byKey nodes are considered")
+      handleLookupWithInternal(lookup, keyInput)
+    }
+
+    private def handleLookupWithInternal(
```
Comment (matthiasS-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:226]: Something got messed up. There is another method with the same name below.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:509]: No, I really mean `exercise` here (renaming from choice though). If they key was brought in scope before (that’s what the rest of the sentence says) then an exercise drops it from the map even if that exercise was not by key.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala (hunk 4 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:222]: as discussed added a description on when Engine queries.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:228]: thx fixed it up

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala (hunk 3 above)
Comment (matthiasS-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/ContractStateMachine.scala:492]: Ok, good to know!

**Review by matthiasS-da — APPROVED** [review]
LGTM

**Review by andreaslochbihler-da — APPROVED** [review]
Thanks a lot for improving the docs.

**Pull request merged**